src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/TeredoHelper.cs

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Net.Sockets;
+using System.Threading;
+
+namespace System.Net.NetworkInformation
+{
+    // This class wraps the native API NotifyStableUnicastIpAddressTable.  The native function's behavior is:
+    // 
+    // 1. If the address table is already stable, it returns ERROR_SUCCESS and a Mib table handle that we must free.
+    //    The passed-in callback will never be called, and the cancelHandle is set to NULL.
+    //
+    // 2. If the address table is not stable, it returns ERROR_IO_PENDING.  The table handle is set to NULL,
+    //    and the cancelHandle is set to a valid handle.  The callback will be called (on a native threadpool thread)
+    //    EVERY TIME the address table becomes stable until CancelMibChangeNotify2 is called on the cancelHandle
+    //    (via cancelHandle.Dispose()).
+    //
+    // CancelMibChangeNotify2 guarantees that, by the time it returns, all calls to the callback will be complete
+    // and that no new calls to the callback will be issued.
+    //
+    // The major concerns of the class are: 1) making sure none of the managed objects needed to handle a native
+    // callback are GC'd before the callback, and 2) making sure no native callbacks will try to call into an unloaded
+    // AppDomain.
+
+    internal class TeredoHelper
+    {
+        // Holds a list of all pending calls to NotifyStableUnicastIpAddressTable.  Also used as a lock to protect its
+        // contents.
+        private static List<TeredoHelper> s_pendingNotifications;
+
+        private readonly Action<object> _callback;
+        private readonly object _state;
+
+        private bool _runCallbackCalled;
+
+        // We explicitly keep a copy of this to prevent it from getting GC'd.
+        private readonly Interop.IpHlpApi.StableUnicastIpAddressTableDelegate _onStabilizedDelegate;
+
+        // Used to cancel notification after receiving the first callback, or when the AppDomain is going down.
+        private SafeCancelMibChangeNotify _cancelHandle;
+
+        static TeredoHelper()
+        {
+            s_pendingNotifications = new List<TeredoHelper>();
```

Comment:
If you just move the initialization of the list to be part of the field declaration, you could get rid of the static cctor and the  type would become beforefieldinit.